edge-modules/edge-dashboard/src/state.rs

```diff
@@ -0,0 +1,108 @@
+use std::fs;
+use std::io::Result;
+use std::env;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum State {
+    Manual,
+    NotProvisioned,
+    NotInstalled,
+    // DPS
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Device {
+    state: State,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    hub_name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    device_id: Option<String>,
+}
+
+impl Device {
+    // makes a new Device struct based on the state
+    pub fn new(state: &str, con_string: String) -> Device {
+        match state {
+            "manual"          => Device::manual_rep(con_string),
+            "not provisioned" => Device::default(State::NotProvisioned),
+            // "dps"          => ,
+            _                 => Device::default(State::NotInstalled),
+        }
+    }
+
+    // returns the manual state representation of device
+    pub fn manual_rep(con_string: String) -> Device {
+        if let Some((hub_name, device_id)) = get_device_details(con_string) {
+            Device {
+                state: State::Manual,
+                hub_name: Some(hub_name),
+                device_id: Some(device_id),
+            }
+        } else {
+            Device::default(State::Manual)
+        }
+    }
+
+    // returns a Device of specified state with all other fields as None
+    pub fn default(state: State) -> Device {
+        Device {
+            state: state,
+            hub_name: None,
+            device_id: None,
+        }
+    }
+}
+
+// returns the connection string given the config.yaml file contents
+pub fn get_connection_string(contents: String) -> Option<String> {
```

Comment:
Instead of doing string pattern matching, I'd suggest parsing the config.yaml into Rust structures. `iotedged` does this already and we should be able to reuse that code. Start by looking at the `init_common` function in `edgelet/iotedged/src/app.rs`.

Reply:
https://github.com/mehcode/config-rs